ethcore/src/engines/authority_round/mod.rs

```diff
@@ -1120,59 +1167,36 @@ impl Engine<EthereumMachine> for AuthorityRound {
 			}
 
 		} else {
-			// Report skipped primaries.
-			if let (true, Some(me)) = (step > parent_step + 1, self.signer.read().address()) {
-				debug!(target: "engine", "Author {} built block with step gap. current step: {}, parent step: {}",
-					   header.author(), step, parent_step);
-				let mut reported = HashSet::new();
-				for s in parent_step + 1..step {
-					let skipped_primary = step_proposer(&*self.validators, &parent.hash(), s);
-					// Do not report this signer.
-					if skipped_primary != me {
-						self.validators.report_benign(&skipped_primary, set_number, header.number());
-						// Stop reporting once validators start repeating.
-						if !reported.insert(skipped_primary) { break; }
- 					}
- 				}
-			}
+			self.report_skipped(header, step, parent_step, validators, set_number);
 		}
 
 		Ok(())
 	}
 
 	// Check the validators.
 	fn verify_block_external(&self, header: &Header) -> Result<(), Error> {
-		// fetch correct validator set for current epoch, taking into account
-		// finality of previous transitions.
 		let active_set;
-		let validators = if self.immediate_transitions {
-			&*self.validators
+		let (validators, set_number) = if self.immediate_transitions {
```

Comment:
This pattern seems to repeat couple of times, might be worth to extract it to a function?

Reply:
I've refactored part of it into the `epoch_set` function. Unfortunately it's not easy to get the whole pattern into a function since both branches are returning different types (`&ValidatorSet` vs `SimpleList`). It could work by returning `Box<ValidatorSet>` from the function but I can't clone `self.validators`.